ds701_book/26-TimeSeries.qmd

```diff
@@ -0,0 +1,953 @@
+  - Provides smooth and adaptable trend and seasonal components.
+
+See also [@hyndman2021forecasting].
+
+## Practical example using Python
+
+Let's demonstrate a practical example of time series decomposition using Python.
+
+**Steps**:
+
+1. Load a sample time series dataset.
+2. Apply a decomposition technique (e.g., classical decomposition or STL).
+3. Visualize the decomposed components (trend, seasonal, and residual).
+
+```{python}
+#| code-fold: false
+import pandas as pd
+import matplotlib.pyplot as plt
+from statsmodels.tsa.seasonal import seasonal_decompose
+
+# Load a sample time series dataset
+data = pd.read_csv('air_passengers_1949_1960.csv', index_col='Date', parse_dates=True)
+ts = data['Number of Passengers']
+```
+
+## `statsmodels` Classical Decomposition -- Additive Model
+
+We'll apply the classical decomposition using the `statsmodels` 
+`seasonal_decompose` function with the additive model.
+
+```{python}
+# Apply classical decomposition
+decomposition = seasonal_decompose(ts, model='additive')
+
+# Plot the decomposed components
+fig, (ax1, ax2, ax3, ax4) = plt.subplots(4, 1, figsize=(8, 8))
+decomposition.observed.plot(ax=ax1)
+ax1.set_ylabel('Observed')
+ax1.set_title('Air Passengers 1949-1960')
+decomposition.trend.plot(ax=ax2)
+ax2.set_ylabel('Trend')
+decomposition.seasonal.plot(ax=ax3)
+ax3.set_ylabel('Seasonal')
+decomposition.resid.plot(ax=ax4)
+ax4.set_ylabel('Residual')
+plt.tight_layout()
+plt.show()
+```
+
+## `statsmodels` Classical Decomposition -- Multiplicative Model
+
+Let's try the multiplicative model:
+
+```{python}
+
+# Apply classical decomposition
+decomposition = seasonal_decompose(ts, model='multiplicative')
+
+# Plot the decomposed components
+fig, (ax1, ax2, ax3, ax4) = plt.subplots(4, 1, figsize=(8, 8))
+decomposition.observed.plot(ax=ax1)
+ax1.set_ylabel('Observed')
+decomposition.trend.plot(ax=ax2)
+ax2.set_ylabel('Trend')
+decomposition.seasonal.plot(ax=ax3)
+ax3.set_ylabel('Seasonal')
+decomposition.resid.plot(ax=ax4)
+ax4.set_ylabel('Residual')
+plt.tight_layout()
+plt.show()
+```
+
+
+## STL Decomposition
+
+The STL decomposition is more flexible and can handle the seasonality in the data
+better.
+
+Let's try the STL decomposition:
+
+```{python}
+from statsmodels.tsa.seasonal import STL
+
+stl = STL(ts, period=12, robust=True)
+result = stl.fit()
+# result.plot()
+
+# Plot the decomposed components
+fig, (ax1, ax2, ax3, ax4) = plt.subplots(4, 1, figsize=(8, 6))
+result.observed.plot(ax=ax1)
+ax1.set_ylabel('Observed')
+ax1.set_title('Air Passengers 1949-1960 (STL)')
+result.trend.plot(ax=ax2)
+ax2.set_ylabel('Trend')
+result.seasonal.plot(ax=ax3)
+ax3.set_ylabel('Seasonal')
+result.resid.plot(ax=ax4)
+ax4.set_ylabel('Residual')
+plt.tight_layout()
+plt.show()
+```
+
+
+
+# Stationarity and Differencing
+
+## Definition of stationarity
+
+A time series is said to be stationary if its statistical properties such as mean, variance, and autocorrelation are constant over time.
```

Comment:
Should we add a definition of autocorrelation somewhere?

Reply:
Good point. I added a few slides and example to define it.